beacon/beacon-common.js

```diff
@@ -0,0 +1,353 @@
+                    });
+
+                    test.done();
+                });
+
+                // Continue polling until all of the results come in.
+                if (session.doneCount < session.sentCount) {
+                    // testharness.js frowns upon the use of explicit timeouts, but there is no way
+                    // around the need to poll for these tests, and there is no use spamming the server
+                    // with requestAnimationFrame() just to avoid the use of step_timeout.
+                    step_timeout(waitForResults.bind(this, session), 100);
+                }
+            }).catch(function(error) {
+                failSession(session, `unexpected error reading response, error='${error}'`);
+            });
+        }
+    );
+}
+
+// Fails all of the tests in the session, meant to be called when an infrastructural
+// issue prevents us from deterministically completing the individual tests.
+function failSession(session, reason) {
+    session.testCases.forEach(function(testCase) {
+        var test = testCase.test;
+        test.unreached_func(reason)();
+    });
+}
+
+// Creates an iframe on the document's body and runs the sample tests from the iframe.
+// The iframe is navigated immediately after it sends the data, and the window verifies
+// that the data is still successfully sent.
+//    funcName: "beacon" to send the data via navigator.sendBeacon(),
+//              "fetch" to send the data via fetch() with the keepalive flag.
+function runSendInIframeAndNavigateTests(funcName) {
+    var iframe = document.createElement("iframe");
+    iframe.id = "iframe";
+    iframe.onload = function() {
+        var tests = Array();
+
+        // Implement the self.buildId extension to identify the parameterized
+        // test in the report.
+        self.buildId = function(baseId) {
+            return `${baseId}-${funcName}-NAVIGATE`;
+        };
+
+        window.onmessage = function(e) {
+            // The iframe will execute sendData() for us and return the result.
+            var testCase = tests[e.data];
+            continueAfterSendingBeacon(true /* sendResult */, testCase);
+        };
+
+        // Implement the self.sendFunc extension to send the beacon indirectly,
+        // from an iFrame that we can then navigate.
+        self.sendFunc = function(testCase) {
+            var iframeWindow = document.getElementById("iframe").contentWindow;
+            // We run into problems passing the testCase over the document boundary,
+            // because of structured cloning constraints. Instead we'll send over the
+            // test case id, and the iFrame can load the static test case by including
+            // beacon-common.js.
+            tests[testCase.origId] = testCase;
+            iframeWindow.postMessage([testCase.origId, testCase.url, funcName], "*");
+        };
+
+        runTests(sampleTests);
```

Comment:
Running this in WebKit & Blink, I see runTests() getting called twice (so we run all 7 sample tests twice, a total of 14 tests). This seems to indicate that the onload handler is called twice. Isn't it the expected behavior that unload gets called twice given that the iframe navigates itself to Microsoft.com at some point?